MergeObjects

danieldresser-ie opened the pull request:
OK, this should be in a pretty decent place to review ... all the weird corner case behaviours seem to be working OK.

There are some odds-n-ends commits here, from back when I was hoping this could share more functionality with BranchCreator and I was looking at BranchCreator closely.

Oh, and a TODO on the weird thing on SceneNode where it uses `h` as the initial value, but then also appends onto h afterwards, I should clean that up.

Anyway, we can't merge this yet because we don't have a tolerable implementation of merge merging yet - the only thing I want you to look at is MergeObjects, and I think that's ready.

johnhaddon reviewed (commented):
Thanks Daniel!

It's going to take me a long time to absorb all this - as you said, it's not simple! Just posting a few preliminary comments so you have a couple of things to consider today. Hopefully I'll be able to get my head into it a bit further on Monday. In the meantime it probably makes sense to work on the merge algo...

Cheers...
John

johnhaddon commented:
Just noting that https://github.com/GafferHQ/gaffer/issues/956 can be closed once this is merged.

danieldresser-ie commented on src/GafferScene/Constraint.cpp:
```diff
@@ -1172,6 +1172,13 @@ Imath::M44f Constraint::computeConstrainedTransform( const Gaffer::Context *cont
 		return inputTransform;
 	}
 
+	// \todo - It's not obvious how we can use something like the hypothetical SceneAlgo::relativeTransform
```
That would be incredibly convenient if true, but existing constraints include things like PointConstraints "xEnabled/yEnabled/zEnabled" toggles, which operate in the space passed to computeConstraint ... so I don't think it's going to be easy?

johnhaddon replied:
Hmm, good point. Is PointConstraint the only one? With another quick scan over the others I didn't see anything else that might be problematic.

danieldresser-ie commented on include/GafferScene/MergeObjects.h:
```diff
@@ -0,0 +1,145 @@
+/// locations. Appropriate for nodes which merge primitives, or convert
+/// transforms to points.
+///
+/// More complicated then you would expect, because it supports arbitrary
```
It's true that it doesn't matter if the implementation was complicated here, but I think it is relevant to programmers using this that the behaviour is complicated - I wouldn't have seen the name "MergeObjects", and expected that it's functionality was a super-set of an "uber-rename" that can assign arbitrary hierarchy locations to each input location.

I've moved this comment to the end of the paragraph, and tried to remove anything non-essential.

danieldresser-ie commented on src/GafferScene/MergeObjects.cpp:
```diff
@@ -0,0 +1,1448 @@
+	// Find the bound from children of this location, either in the existing scene, or added
+	// by merges targetting child locations as destinations.
+
+	// To know if we need to check the child bounds, we need to know if there are any destinations below this
+	// location. We can efficiently query this from the mergeLocationData.
+	ConstMergeLocationDataPtr mergeLocationData = IECore::runTimeCast< const MergeLocationData >( mergeLocationPlug()->getValue() );
```
I'm sure it does have some sort of performance implications. In a simplistic view of read performance, grabbing mergeLocations plug from the cache is one map lookup ( of an admittedly large map ), whereas finding the location in branchesPlug() requires a map lookup for each level of the hierarchy, and then you'd have to do another plug evaluation of the filter to check if anything is pruned ( which is also unique at each location ). But there probably are cases where the hash pressure is a much bigger problem than doing some map lookups. Probably could be reasonable to add some more code paths to check the tree data and the filter, and skip access to mergeLocationPlug when not needed.

johnhaddon replied:
> In a simplistic view of read performance, grabbing mergeLocations plug from the cache is one map lookup

Not sure I've read this right, but it seems like you might be implying that most `__mergeLocation` evaluations are cache hits? But it looks like `MergeLocationsData::hash()` includes `path` in the hash, meaning that we're computing something at `__mergeLocation` every time, even when there's no merging going on at that location. That's what led me to think this might be a fair bit worse than doing an initial lookup on the TreeData (which _is_ cached).

I suggest we park this temporarily while focussing on getting a final `MeshAlgo::merge()` implementation sorted. But I do think it'd be worth doing a bit of further investigation before we release.

danieldresser-ie replied:
I probably should have documented my thoughts more before moving on, took me a bit to remember this, but anyway:

Yes, we hit MergeLocations for every location, including locations where no merging is going on. But the work performed in locations where there is no merging happening is the same work we would need to perform anyway in order to check if we should access MergeLocations.

So the trade-off is just whether we cache this work between different locations accessing MergeLocations - caching it is a minor speedup, but consumes more cache.


I've added two new tests in [5d5ccc3](https://github.com/GafferHQ/gaffer/pull/5984/commits/5d5ccc37b53cfdd55b0c45107475df94ff334948)

`testPerformanceUnprocessedLocations` tests the case where we aren't processing most locations, and MergeLocationData is usually NullObject anyway, which gives us the best case of seeing a benefit from not bothering to access it.

`testPerformanceDeepPaths` tests the case where there are deep paths being accessed repeatedly - the children of a location will access MergeLocationData in order to check their parent, so there are lots of accesses to every location, and querying the tree is expensive, giving maximum benefit to caching MergeLocationData.

[19f5fdb](https://github.com/GafferHQ/gaffer/pull/5984/commits/19f5fdb8a5483e28a10ae2a5762f281537a740ba) adds some extra code to perform the extra tree tree query and skip accessing MergeLocationsData whenever possible.

This causes testPerformanceDeepPaths to get 10% slower, and testPerformanceUnprocessedLocations to get 5% faster.

There is probably an argument though that testPerformanceUnprocessedLocations is more plausible to occur in production scenes, and while the performance benefit isn't huge, it does reduce cache usage by 50% in this extreme case - so maybe it's worth it?

If you want to go with 5d5ccc3, I should clean it up tomorrow. If you don't think it's needed, it would be fine to just drop the last commit.

johnhaddon replied:
I hadn't anticipated the need for the filter checks in `accessMergeLocation()` - with my naive BranchCreator-primed hat on, I think I assumed the tree had all the info we need. But it doesn't, because that only stores destinations, right, and things that need pruning aren't stored?

For what it's worth, I'm seeing a 10% improvement in `testPerformanceUnprocessedLocations` and no real change in `testPerformanceDeepPaths` here on my workstation (32 core Xeon). My general feeling is that the merging-a-small-number-of-locations-in-a-large-scene is most likely our most important use case, but we also have to balance that against the maintainability of the code. I'll leave it to your judgement...

johnhaddon replied:
Alternative thought : is it possible to tighten up the hash for `mergeLocationPlug()`, so that we do fewer computes on that plug by omitting `path` whenever we can?

danieldresser-ie commented on src/GafferScene/MergeMeshes.cpp:
```diff
@@ -0,0 +1,128 @@
+			continue;
+		}
+
+		const IECoreScene::MeshPrimitive * m = IECore::runTimeCast< const IECoreScene::MeshPrimitive >( sources[i].first.get() );
+		if( !m )
+		{
+			// \todo - we don't really want this logic to depend on the source paths, so it's nice to not
+			// have them passed in ... but that prevents us from throwing a useful exception here.
+			// Actually kinda feels like the most elegant approach would be a custom exception type,
+			// like MergeObjectException( "Source is not mesh", i ), and then MergeObjects would translate
+			// that to a regular exception with the source path included. We don't really do that elsewhere
+			// though, would that be weird?
+			throw IECore::Exception(
+				fmt::format(
+					"Source must be mesh, got {}.", sources[i].first->typeName()
+				)
+			);
+		}
```
I'm not generally a fan of silently failing / discarding things, but it does definitely make things simpler here, I'll just do that.

johnhaddon replied:
I think ignoring non-meshes is pretty intuitive for a node called MergeMeshes, and it matches established behaviour in nodes like MeshTangents and MeshToLevelSet.

danieldresser-ie commented on include/GafferScene/Private/ChildNamesMap.h:
```diff
@@ -75,6 +76,8 @@ class ChildNamesMap : public IECore::Data
 		/// Combines multiple input sets, accounting for the name remapping.
 		IECore::PathMatcher set( const std::vector<IECore::ConstPathMatcherDataPtr> &inputSets ) const;
 
+		static void makeNameUnique( IECore::InternedString &name, const std::unordered_set<IECore::InternedString> &existingNames );
```
Addressed.

danieldresser-ie commented on include/GafferScene/MergeObjects.h:
```diff
@@ -0,0 +1,145 @@
+/// locations, creating the destination if it doesn't exist already. Then
+/// all source locations that are not also a destination are pruned.
+///
+/// Derived classe just need to implement `mergeObjects()` to do the actual
```
Addressed.

danieldresser-ie commented on include/GafferScene/MergeObjects.h:
```diff
@@ -0,0 +1,145 @@
+		/// @name Actual object merge function
+		/// This must be implemented by derived classes. It receives a vector of pairs of objects
+		/// and their transforms.
+		///
```
Addressed.

danieldresser-ie commented on python/GafferSceneUI/MergeObjectsUI.py:
```diff
@@ -0,0 +1,96 @@
+			"""
+			The filter used to control which parts of the scene are
+			processed. A Filter node should be connected here.
+			Any filtered locations which aren't used as part of a destination will be pruned.
```
Used the suggestion.

danieldresser-ie commented on python/GafferSceneUI/MergeObjectsUI.py:
```diff
@@ -0,0 +1,96 @@
+			will be preserved except for the newly added merge locations.
+			"""
+
+		]
```
Used the suggestion.

danieldresser-ie commented on src/GafferScene/MergeObjects.cpp:
```diff
@@ -0,0 +1,1448 @@
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new ScenePlug( "source" ) );
+
+	// TODO - default should be constructor argument
```
Addressed.

johnhaddon replied:
We also need to remove the `// TODO - default should be constructor argument` comment.

johnhaddon replied:
This TODO still needs removing.

danieldresser-ie commented on src/GafferScene/MergeObjects.cpp:
```diff
@@ -0,0 +1,1448 @@
+			std::unordered_map< InternedString, SourcePaths > destinations;
+		};
+
+		TreeData( const ScenePlug *inPlug, const FilterPlug *filterPlug, const StringPlug *destinationPlug, const Context *context )
```
Addressed.

danieldresser-ie commented on src/GafferScene/MergeObjects.cpp:
```diff
@@ -0,0 +1,1448 @@
+			;
+		}
+
+		static void hash( const ScenePlug *inPlug, const FilterPlug *filterPlug, const StringPlug *destinationPlug, const Context *context, IECore::MurmurHash &h )
```
Addressed.

danieldresser-ie commented on src/GafferScene/MergeObjects.cpp:
```diff
@@ -0,0 +1,1448 @@
+		TreeData( const ScenePlug *inPlug, const FilterPlug *filterPlug, const StringPlug *destinationPlug, const Context *context )
+			:	m_root( new Location() )
+		{
+			auto f = [this, destinationPlug]( const GafferScene::ScenePlug *scene, const GafferScene::ScenePlug::ScenePath &path )
```
Addressed.

danieldresser-ie commented on src/GafferScene/MergeObjects.cpp:
```diff
@@ -0,0 +1,1448 @@
+			destinationPlug->hash( h );
+		}
+
+		void addBranch( const StringPlug *destinationPlug, const ScenePlug::ScenePath &path )
```
Addressed.

danieldresser-ie commented on src/GafferScene/MergeObjects.cpp:
```diff
@@ -0,0 +1,1448 @@
+{
+
+	public :
+		MergeLocationData( const TreeData::Location *location, const ScenePlug::ScenePath &path, const MergeObjects *branchCreator, const Context *context, const int rootFilterValue )
```
Addressed.

danieldresser-ie commented on src/GafferScene/MergeMeshes.cpp:
```diff
@@ -0,0 +1,128 @@
+{
+	std::vector< IECoreScene::ConstMeshPrimitivePtr > meshes;
+	std::vector< const IECoreScene::MeshPrimitive * > meshPointers;
+	for( unsigned int i = 0; i < sources.size(); i++ )
```
Addressed.

danieldresser-ie commented:
Addressed first round feedback. Was unsure whether to squash or not, but it all seemed straightforward enough that it made sense to squash.

johnhaddon commented:
Thanks for the update Daniel - I've been through and resolved or replied to the previous discussions. One thing I forgot to mention the first time around - I'd like to omit the BranchCreator commits from this PR. They're not really related, and I'm not sure how much benefit they really have - feel free to open a separate PR for them if you want though.

danieldresser-ie commented:
This is in a pretty sloppy state at the moment - it's rebased on top of #6026 which is full of patch commits, and I haven't yet trimmed out the BranchCreator commits.

But this should be good enough for doing interactive testing of #6026 as you requested.

johnhaddon commented:
Spent an hour or so messing about with this this afternoon, and it all worked very nicely. Every time I thought I'd found a bug it turned out just to be a side-effect of the zero-initialisation of values when one mesh doesn't have a primvar that another mesh does have. This happened for me both with `Cs` and `N`, in both cases making the mesh black in the viewport. I think this might warrant a note in the documentation for the node, and perhaps in the case of `N`, a warning at runtime?

danieldresser-ie commented:
Re-rebased onto #6026
Split off #6038 and #6039 
Added note about missing normals to node doc.

johnhaddon commented:
Thanks Daniel - could you rebase this one more time please, now that https://github.com/GafferHQ/gaffer/pull/6026 is merged. We also have a [parked conversation](https://github.com/GafferHQ/gaffer/pull/5984#discussion_r1706706647) that is probably worth continuing before finally merging this.

danieldresser-ie commented:
While waiting for feedback on the optimization question, I've added MergePoints and MergeCurves.

Also added [a82ce06](https://github.com/GafferHQ/gaffer/pull/5984/commits/a82ce06c0af27cbd0dc680dae48c5f46dc5145a7) with special case handling of "type" on Curves, since it won't be picked up by the renderer backends if we promote it from Constant to Vertex.

johnhaddon reviewed (commented):
Thanks Daniel! I've replied to our previous conversation about performance, noted a TODO from last time, and added a couple of comments for the latest commits. Hopefully we're pretty close now.
Cheers...
John

danieldresser-ie commented on startup/gui/menus.py:
```diff
@@ -281,6 +281,8 @@ def __lightCreator( nodeName, shaderName, shape ) :
 nodeMenu.append( "/Scene/Object/Mesh Split", GafferScene.MeshSplit, searchText = "MeshSplit" )
 nodeMenu.append( "/Scene/Object/Merge Meshes", GafferScene.MergeMeshes, searchText = "MergeMeshes" )
+nodeMenu.append( "/Scene/Object/Merge Meshes", GafferScene.MergePoints, searchText = "MergePoints" )
+nodeMenu.append( "/Scene/Object/Merge Meshes", GafferScene.MergeCurves, searchText = "MergeCurves" )
```
Fix squashed into [6ecbb07](https://github.com/GafferHQ/gaffer/pull/5984/commits/6ecbb078f31adeabbc24c0297a7aca7f560fa29f)

danieldresser-ie commented on src/GafferScene/MergePoints.cpp:
```diff
@@ -0,0 +1,92 @@
+		[&]() {
+			return IECoreScenePreview::PrimitiveAlgo::mergePrimitives( points, context->canceller() );
+		}
+	);
```
Addressed in [41f3711](https://github.com/GafferHQ/gaffer/pull/5984/commits/41f371185bd8fa0a7f883c87917a44dd155d0af2)

danieldresser-ie commented:
I've removed the outdated TODO, and gone with your suggestion to spend the upfront evaluation of `__tree` in order to give the more precise hash for __mergeLocation. The hash for objectPlug was evaluating __tree anyway, and going with this approach is simpler, and gives the performance benefit of the previous optimization commit, while avoiding the downside in testPerformanceDeepPaths.

If this looks good, it just needs a squash before merging.

johnhaddon commented:
Thanks Daniel! Squashed and merging...
